contracts/Marketplace.sol

```diff
@@ -1,74 +1,272 @@
-import "openzeppelin-solidity/contracts/lifecycle/Pausable.sol";
-
-contract Service is Ownable, Pausable {
-    bytes32 public sid;
-    bytes32 public latest;
-    mapping (bytes32 => Version) public versions;
-    mapping (bytes32 => mapping(address => bool)) private _versionAccess;
-
-    struct Version {
-        bytes32 id;
-        bytes32 location;
-        uint256 price;
-        bool _exists;
-    }
+pragma solidity >=0.5.0 <0.6.0;
 
-    event VersionCreated(bytes32 sid, bytes32 id, bytes32 location, uint256 price);
+import "../node_modules/openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "../node_modules/openzeppelin-solidity/contracts/lifecycle/Pausable.sol";
 
-    constructor(address creator, bytes32 _sid) Ownable() Pausable() public {
-        sid = _sid;
-        transferOwnership(creator);
-        addPauser(creator);
-        renouncePauser();
-    }
+contract Marketplace is Ownable, Pausable {
+  // ------------------------------------------------------
+  // Structures
+  // ------------------------------------------------------
 
-    function createVersion(bytes32 versionHash, bytes32 location, uint256 price) public onlyOwner whenNotPaused {
-        require(!versions[versionHash]._exists, "This version already exists");
-        versions[versionHash] = Version({
-            id: versionHash,
-            location: location,
-            price: price,
-            _exists: true
-        });
-        latest = versionHash;
-        emit VersionCreated(sid, versionHash, location, price);
-        return;
-    }
+  struct Service {
+    address owner;
+    string sid;
+    Version[] versions;
+    Offer[] offers;
+    Payment[] payments;
+  }
+
+  struct Version {
+    string hash;
+    string url;
+  }
+
+  struct Offer {
+    uint price;
+    // address payment; // TODO: to implement later
+    address payable seller;
+    bool active;
+  }
+
+  struct Payment {
+    uint offerIndex;
+    address purchaser;
+  }
+
+  // ------------------------------------------------------
+  // TODO: Events
+  // ------------------------------------------------------
+
+  // ------------------------------------------------------
+  // State variables
+  // ------------------------------------------------------
+
+  Service[] public services; //TODO: shouldn't it be private?
+
+  // ------------------------------------------------------
+  // Constructor
+  // ------------------------------------------------------
+
+  constructor() public {}
 
-    function requestAccess(bytes32 version) public whenNotPaused {
-        require(!_versionAccess[version][msg.sender], "You already have access to this version of the service");
-        _versionAccess[version][msg.sender] = true;
-        return;
+  // ------------------------------------------------------
+  // Check functions
+  // ------------------------------------------------------
+
+  function checkServiceOwner(Service memory service) private view {
```

Comment:
This should be modifier i guess but need to check

Reply:
I don't think i can do a modifier with this but that's will be way better.
I need the service index in order to check ownership. Can we pass parameter to modifier?